python_modules/dagster/dagster/_core/storage/upath_io_manager.py

```diff
@@ -0,0 +1,195 @@
+    """
+    Abstract IOManager base class compatible with local and cloud storage via `fsspec` (using `universal-pathlib`).
+    What it handles for the use:
+     - working with any filesystem supported by `fsspec`
+     - handling loading multiple upstream asset partitions via PartitionMapping
+     (returns a dictionary with mappings from partition_keys to loaded objects)
+     - the `get_metadata` method can be customized to add additional metadata to the outputs
+     - the `allow_missing_partitions: bool` metadata value can control
+     either raising an error or skipping on missing partitions
+    """
+
+    extension: str = ""  # override in child class
+
+    @abstractmethod
+    def dump_to_path(self, context: OutputContext, obj: Any, path: UPath):
+        """
+        Child classes should override this method to write the object to the filesystem.
+
+        Args:
+            context:
+            obj:
+            path:
+
+        Returns:
+
+        """
+        ...
+
+    @abstractmethod
+    def load_from_path(self, context: InputContext, path: UPath) -> Any:
+        """
+        Child classes should override this method to load the object from the filesystem.
+
+        Args:
+            context:
+            path:
+
+        Returns:
+
+        """
+        ...
+
+    @staticmethod
+    def get_metadata(context: OutputContext, obj: Any) -> Dict[str, MetadataValue]:
+        """
+        Child classes should override this method to add custom metadata to the outputs.
+
+        Args:
+            context:
+            obj:
+
+        Returns:
+
+        """
+        return {}
+
+    def __init__(
+        self,
+        base_path: UPath,
+    ):
+        assert self.extension == "" or "." in self.extension
+
+        self._base_path = base_path
+
+    def has_output(self, context: OutputContext) -> bool:
+        return self._get_path(context).exists()
+
+    @staticmethod
+    def _has_multiple_partitions(context: Union[InputContext, OutputContext]):
+        key_range = context.asset_partition_key_range
+        return key_range.start != key_range.end
+
+    def _get_path(self, context: Union[InputContext, OutputContext]) -> UPath:
+        if context.has_asset_key:
+            # we are dealing with an asset
+            context_path = ["assets"] + list(context.asset_key.path)
+
+            if context.has_asset_partitions:
+                # add partition
+                context_path += [context.partition_key]
+        else:
+            # we are dealing with an op output
+            context_path = ["runs"] + list(context.get_identifier())
+
+        return self._base_path.joinpath(*context_path).with_suffix(self.extension)
+
+    def load_input(self, context: InputContext) -> Union[Any, Dict[str, Any]]:
+        # In this load_input function, we vary the behavior based on the type of the downstream input
+        # if the type is a List, we load a list of mapped partitions.
+        path = self._get_path(context)
```

Comment:
In the case where there are multiple upstream partitions, it doesn't make sense to get a single path, right? I see that you handle this later by backtracking up to `path.parent`, but I think this is a little bit awkward because it kind of requires "undoing" what we did here. And the logic for constructing the path for a partition is spread between `_get_path` and the `has_multiple_partitions` branch of `load_input`.

A couple possible solutions:
- Allow `_get_path` to return a dictionary mapping partition keys to paths?
- Define a `_get_path_for_partition` method that gets used in the `has_multiple_partitions` branch as well as called internally within `_get_path`?

Reply:
went with (2)